### PR TITLE
refactor automerge to support merge for protected branch

### PR DESCRIPTION
**Pull request by pxLi**
Previous auto-merge workflow only works for auto-merge into an unprotected branch, as how cudf's automerge does.
But this is not reasonable when we switch `branch-0.3` as default branch, so I added auto-approve for this automation workflow, and it will try approving auto-merged pr if possible.

demo PR: https://github.com/pxLi/spark-rapids/pull/153

NOTE: a new secret `NVAUTO_TOKEN` has to be added to repo since github requires a different account to approve the PR. `nvauto` is our automation account which has already been invited to spark-rapids repo.

### CODE REVIEW

**Comment by pxLi** [comment]
build

**Comment by pxLi** [comment]
build

**Comment by pxLi** [comment]
build

**Comment by pxLi** [comment]
build

**Comment by pxLi** [comment]
build

File: .github/workflows/auto-merge/automerge
Comment (jlowe) [inline .github/workflows/auto-merge/automerge:134]: Won't this seriously eat into our github action minutes?

Comment (pxLi) [reply .github/workflows/auto-merge/automerge:134]: `github actions` is currently free to use in public repo. And we can easily turn this off or move it back to our jenkins if github change their mind :)
image

**Comment by pxLi** [comment]
> > only works for auto-merge into an unprotected branch, as how cudf's automerge does
> 
> Is that really the case? I find it very surprising that cudf's repo would leave the new development branch unprotected. Also my personal experience with cudf during auto-merge periods is that the new branch still requires CI checks and approvals before developers can merge to the branch, yet GPUtester can still merge the PRs without approval.


yeah, they keep the CI check but added the `[skip ci]` thing for auto-merge scenario to pass the CI checks. I did the same thing for our new blossom-ci pipeline [PR #642](https://github.com/NVIDIA/spark-rapids/pull/642/files#diff-bb1d3d454db208edc1f6f507f39433cfR129-R134)

**Comment by pxLi** [comment]
build

File: .github/workflows/auto-merge/automerge
Comment (jlowe) [inline .github/workflows/auto-merge/automerge:131]: Curious, why was this removed?

File: .github/workflows/auto-merge/automerge
Comment (jlowe) [inline .github/workflows/auto-merge/automerge:89]: Why were the instructions on how to manually resolve the auto-merge removed?

Comment (pxLi) [reply .github/workflows/auto-merge/automerge:89]: we can hardly tell if it is merge conflict or other issues by status code, we have to add some regex for matching in message, but messages could be changed in the future. So I decided to replace that with the original failure info from response. 

And people should know how to fix the conflict I assume :)

File: .github/workflows/auto-merge/automerge
Comment (pxLi) [reply .github/workflows/auto-merge/automerge:131]: try to make it a pure script lol

Comment (jlowe) [reply .github/workflows/auto-merge/automerge:131]: What is gained by this change?  That check doesn't preclude the code from being used like a script, right?  It seems like if anything this change keeps the code from being reusable in other workflows as a module.  If this change didn't fix anything, it makes the code less reusable for no benefit.

Comment (pxLi) [reply .github/workflows/auto-merge/automerge:131]: hmm, actually that script is not reusable before, I just want to unblock the auto-merge stuff earlier. 
As you request, let me refactor the whole thing. Give me some minutes~

**Comment by pxLi** [comment]
build

File: .github/workflows/auto-merge/automerge
Comment (jlowe) [reply .github/workflows/auto-merge/automerge:131]: Yes, you're right, it needs a bit more work than just the check.  Can clean this up in a followup.

**Review by jlowe — APPROVED** [review]

**Pull request merged**
